### PR TITLE
feat patch unhealthy gpu list to node annotations

### PR DESCRIPTION
**Pull request by WingkaiHo**
Signed-off-by: yongjiahe <yongjiahe@tuputech.com>

Patch unhealthy gpu list to node annotations, help scheduler to exclude unhealthy gpu index

### CODE REVIEW

File: pkg/plugin/nvidia/kube_interactor.go [hunk 1]
```diff
@@ -114,3 +116,42 @@ func (ki *KubeInteractor) PatchGPUResourceOnNode(gpuCount int) error {
...
+
+	if len(unhealthyGPUs) > 0 {
+		unhealthyGPUsStr = strings.Join(unhealthyGPUs, ",")
+	}
+
+	err = wait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+		var node *v1.Node
+		node, err = ki.clientset.CoreV1().Nodes().Get(context.TODO(), ki.nodeName, metav1.GetOptions{})
+		if err != nil {
+			klog.Info("failed to get node %s: %v", ki.nodeName, err)
```
Comment (Thor-wl) [inline pkg/plugin/nvidia/kube_interactor.go]: Please add a reasonable log level all through the PR, too.

Comment (WingkaiHo) [reply pkg/plugin/nvidia/kube_interactor.go]: OK, fix

File: pkg/plugin/nvidia/nvidia.go [hunk 2]
```diff
@@ -119,13 +132,23 @@ func checkHealth(stop <-chan struct{}, devices []*Device, unhealthy chan<- *Devi
 
 		e, err := nvml.WaitForEvent(eventSet, 5000)
 		if err != nil && e.Etype != nvml.XidCriticalError {
+			if firstTime {
```
Comment (william-wang) [inline pkg/plugin/nvidia/nvidia.go:135]: Would you describe what is your case?  Why we have to patch unhealthy for the first time?

Comment (WingkaiHo) [reply pkg/plugin/nvidia/nvidia.go:135]: First time to patch use to clear node annotations if all device healthy. Such as the user restart node and fix gpu error.

File: pkg/plugin/nvidia/server.go [hunk 3]
```diff
@@ -255,9 +255,16 @@ func (m *NvidiaDevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.Device
...
 			// FIXME: there is no way to recover from the Unhealthy state.
+			isChange := false
+			if d.Health != pluginapi.Unhealthy {
+				isChange = true
+			}
 			d.Health = pluginapi.Unhealthy
 			log.Printf("'%s' device marked unhealthy: %s", m.resourceName, d.ID)
 			s.Send(&pluginapi.ListAndWatchResponse{Devices: m.virtualDevices})
+			if isChange {
+				m.kubeInteractor.PatchUnhealthyGPUListOnNode(m.physicalDevices)
```
Comment (WingkaiHo) [inline pkg/plugin/nvidia/server.go:266]: When gpu device status frist time from healthy to unhealthy, it patch unhealth devices list annotations to k8s api server.  It make less concurrency to visit k8s api server

File: pkg/plugin/nvidia/nvidia.go (hunk 2 above)
Comment (WingkaiHo) [reply pkg/plugin/nvidia/nvidia.go:135]: It use to reset node annotations if all device healthy, when device plugin restart.

Comment (WingkaiHo) [reply pkg/plugin/nvidia/nvidia.go:135]: @william-wang

**Review by william-wang — APPROVED** [review]
/lgtm
/approve

**Pull request merged**